File: typescript-solution/src/sort.ts
export type Stack = "STANDARD" | "SPECIAL" | "REJECTED";

const VOLUME_BULKY_THRESHOLD_CM3 = 1000000;
const DIMENSION_BULKY_THRESHOLD_CM = 150;
const HEAVY_THRESHOLD_KG = 20;

function ensureFiniteNonNegative(value: number, name: string): void {
  if (!Number.isFinite(value)) {
    throw new TypeError(`${name} must be a finite number`);
  }
  if (value < 0) {
    throw new RangeError(`${name} must be >= 0`);
  }
}

export function sort(
  width: number,
  height: number,
  length: number,
  mass: number
): Stack {
  ensureFiniteNonNegative(width, "width");
  ensureFiniteNonNegative(height, "height");
  ensureFiniteNonNegative(length, "length");
  ensureFiniteNonNegative(mass, "mass");

  const volume = width * height * length;
  const bulky =
    volume >= VOLUME_BULKY_THRESHOLD_CM3 ||
    width >= DIMENSION_BULKY_THRESHOLD_CM ||
    height >= DIMENSION_BULKY_THRESHOLD_CM ||
    length >= DIMENSION_BULKY_THRESHOLD_CM;
  const heavy = mass >= HEAVY_THRESHOLD_KG;

  if (bulky && heavy) return "REJECTED";
  if (bulky || heavy) return "SPECIAL";
  return "STANDARD";
}

File: python-solution/src/sort.py
from __future__ import annotations

import math
from typing import Literal, Union

Stack = Literal["STANDARD", "SPECIAL", "REJECTED"]
Number = Union[int, float]

VOLUME_BULKY_THRESHOLD_CM3 = 1000000
DIMENSION_BULKY_THRESHOLD_CM = 150
HEAVY_THRESHOLD_KG = 20


def _ensure_finite_non_negative(value: Number, name: str) -> None:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise TypeError(f"{name} must be an int or float")
    if not math.isfinite(value):
        raise TypeError(f"{name} must be finite")
    if value < 0:
        raise ValueError(f"{name} must be >= 0")


def sort(width: Number, height: Number, length: Number, mass: Number) -> Stack:
    """
    Dispatch a package into STANDARD, SPECIAL, or REJECTED based on its dimensions and mass.

    Units:
      - width/height/length: centimeters
      - mass: kilograms
    """

    _ensure_finite_non_negative(width, "width")
    _ensure_finite_non_negative(height, "height")
    _ensure_finite_non_negative(length, "length")
    _ensure_finite_non_negative(mass, "mass")

    volume = width * height * length
    bulky = (
        volume >= VOLUME_BULKY_THRESHOLD_CM3
        or width >= DIMENSION_BULKY_THRESHOLD_CM
        or height >= DIMENSION_BULKY_THRESHOLD_CM
        or length >= DIMENSION_BULKY_THRESHOLD_CM
    )
    heavy = mass >= HEAVY_THRESHOLD_KG

    if bulky and heavy:
        return "REJECTED"
    if bulky or heavy:
        return "SPECIAL"
    return "STANDARD"

